crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -1141,6 +1045,390 @@ impl<'a, 'o> TimelineEventHandler<'a, 'o> {
+                                },
+                            ) {
+                                // There's a previous event with the same date: remove the divider.
+                                trace!("removed day divider @ {i} between two events that have the same date");
+                                ops.push(DayDividerOperation::Remove(i - 1));
+                            } else {
+                                // There's no previous event or there's one with a different date:
+                                // replace the current divider.
+                                trace!("replacing day divider @ {i} with new timestamp from event");
+                                ops.push(DayDividerOperation::Replace(i - 1, ts));
+                            }
+                        }
+                    } else if let Some(prev_event) = prev_item.as_event() {
+                        // The event is preceded by another event. If they're not the same date,
+                        // insert a date divider.
+                        let prev_ts = prev_event.timestamp();
+                        if timestamp_to_date(prev_ts) != timestamp_to_date(ts) {
+                            trace!("inserting day divider @ {} between two events with different dates", i);
+                            ops.push(DayDividerOperation::Insert(i, ts));
+                        }
+                    }
+                } else {
+                    // The event was the first item, so there wasn't any day divider before it:
+                    // insert one.
+                    trace!("inserting the first day divider @ {}", i);
+                    ops.push(DayDividerOperation::Insert(i, ts));
+                }
+
+                prev_item = Some((item, ts));
+                last_event_ts = Some(ts);
+            }
+        }
+
+        //#[cfg(debug)]
+        // TODO: in non-debug mode, set to `None` instead.
+        let initial_state = Some(self.items.iter().cloned().collect());
+
+        // Record the deletion offset.
+        let mut offset = 0i64;
+        // Remember what the maximum index was, so we can assert that it's
+        // non-decreasing.
+        let mut max_i = 0;
+
+        for op in &ops {
+            match *op {
+                DayDividerOperation::Insert(i, ts) => {
+                    assert!(i >= max_i);
+
+                    // Keep push semantics, if we're inserting at the front or the back.
+                    let at = (i64::try_from(i).unwrap() + offset)
+                        .min(i64::try_from(self.items.len()).unwrap());
+                    assert!(at >= 0);
+                    let at = at as usize;
+
+                    let item = self.meta.new_timeline_item(VirtualTimelineItem::DayDivider(ts));
+
+                    if at == 0 {
+                        self.items.push_front(item);
+                    } else if at == self.items.len() {
+                        self.items.push_back(item);
```

Comment:
I'm not sure but I think it's:

```suggestion
                    } else if at == self.items.len() - 1 {
                        self.items.push_back(item);
```

Thoughts?

Reply:
No, this would be an insertion *before* the final item.